src/functional/ffront/foast_to_itir.py

```diff
@@ -31,61 +32,80 @@
 
 
-class TypeKind(enum.Enum):
-    FIELD = 0
-    SCALAR = 1
-    UNKNOWN = 2
+def is_local_kind(symbol_type: ct.FieldType) -> bool:
+    assert isinstance(symbol_type, ct.FieldType)
+    if symbol_type.dims == ...:
+        return False
+    return any(dim.kind == DimensionKind.LOCAL for dim in symbol_type.dims)
 
 
-def resulting_type_kind(symbol_type: ct.SymbolType) -> TypeKind:
-    """
-    Determine whether the resulting type kind of ``symbol_type`` is scalar, field or unknown.
-
-    Useful for determining whether an expression of ``symbol_type`` should be treated as a
-    value or iterator expression during lowering.
-
-    Examples:
-    ---------
-    >>> resulting_type_kind(ct.DeferredSymbolType(constraint=None)).name
-    'UNKNOWN'
+class IteratorTypeKind(enum.Enum):
+    VALUE = 0
+    ITERATOR = 1
+    ENCAPSULATED_ITERATOR = 2
 
-    >>> resulting_type_kind(ct.ScalarType(kind=ct.ScalarKind.BOOL)).name
-    'SCALAR'
 
-    >>> resulting_type_kind(ct.TupleType(types=[ct.ScalarType(kind=ct.ScalarKind.FLOAT64)])).name
-    'SCALAR'
+def iterator_type_kind(
+    symbol_type: ct.ScalarType | ct.FieldType | ct.TupleType,
+) -> IteratorTypeKind:
     """
-    match symbol_type:
-        case ct.DeferredSymbolType(constraint=None):
-            return TypeKind.UNKNOWN
-        case ct.TupleType(types=subtypes):
-            if subtypes:
-                return resulting_type_kind(subtypes[0])
-
-    match type_info.type_class(symbol_type):
-        case ct.FieldType:
-            return TypeKind.FIELD
-        case ct.ScalarType:
-            return TypeKind.SCALAR
-
-    return TypeKind.UNKNOWN
-
-
-def can_be_value_or_iterator(symbol_type: ct.SymbolType):
-    return resulting_type_kind(symbol_type) is not TypeKind.UNKNOWN
-
-
-def is_field(expr: foast.Expr) -> bool:
-    return type_info.type_class(expr.type) is ct.FieldType
+    Return the kind of type on iterator level an foast expr of given type corresponds to.
+
+    This function is used both to decide on how to lower an foast expression
+    of the given type and how to handle such expressions in other expressions.
+
+    - VALUE: The expression is a value, e.g. a scalar.
+    - ITERATOR: The expression can be derefed returning a value or composite
+        object of values (e.g. tuple).
+    - ENCAPSULATED_ITERATOR: The expression is a composite object (e.g. tuple)
+        that contains at least one iterator.
+
+    +------------------------------------+------------------------+
+    | Expr                               | Iterator Type Kind     |
+    +====================================+========================+
+    | 1                                  | VALUE                  |
+    | regular_field                      | ITERATOR               |
+    | local_field                        | ITERATOR               |
+    | (1, 1)                             | VALUE                  |
+    | (1, regular_field)                 | ITERATOR               |
```

Comment:
Looking at the definition of `ENCAPSULATED_ITERATOR` above, this seems to meet it. A tuple that contains at least one iterator. Am I reading the definition wrong? Or does it also meet the definition of `ITERATOR` (since it is treated as an iterator of tuples),  and that takes precedence?

Reply:
The description of `VALUE, ITERATOR, ENCAPSULATED_ITERATOR` is on iterator level, which was misleading as it used _expression_ which was then also used as the column heading for FOAST expressions in the table. I changed the descriptions to _lowered expression_ and the heading to _FOAST Expr_. With respect to the example: `(1, regular_field)` will be lowered to an iterator of tuples so it's type kind is `ITERATOR`.

Reply:
Much clearer now, thanks